### PR TITLE
Damianvp/whack a mole

### PR DESCRIPTION
**Pull request by Dam24**
Whack a Mole game task

### CODE REVIEW

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            return false;
+        }
+    }
+
+    /**
+     * Method to verify if game continue.
+     */
+    void verifyGameStatus() {
+        if (this.molesLeft == 0) {
+            System.out.println("Great you win!!!");
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java]: MINOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            while (moles > 0) {
+                int x = random.nextInt(max - min);
+                int y = random.nextInt(max - min);
+                moleIsPlaced = place(x, y);
+                if (moleIsPlaced) {
+                    moles--;
+                }
+            }
+        }
+        System.out.println("Is not possible place more than fdsfsdf");
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java]: MINOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        }
+    }
+
+    /**
+     * Method to verify if game continue.
+     */
+    void verifyGameStatus() {
+        if (this.molesLeft == 0) {
+            System.out.println("Great you win!!!");
+            System.out.println(String.format("Score: %s", this.score));
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java]: MINOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+     */
+    void printGridToUser() {
+        System.out.println(Arrays.deepToString(this.trackerMoleGrid));
+    }
+
+    /**
+     * Method to print the complete grid game status.
+     */
+    void printGrid() {
+        System.out.println(Arrays.deepToString(this.moleGrid));
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java]: MINOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+     * Method to verify if game continue.
+     */
+    void verifyGameStatus() {
+        if (this.molesLeft == 0) {
+            System.out.println("Great you win!!!");
+            System.out.println(String.format("Score: %s", this.score));
+            return;
+        }
+        if (this.attemptsLeft <= 0) {
+            System.out.println("You Loss...  Try Again");
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java]: MINOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    boolean validatedCoordinates(int x, int y) {
+        if (x > INVALID_RANGE && y > INVALID_RANGE && x < this.moleGrid.length && y < this.moleGrid.length) {
+            if (x == -1 && y == -1) {
+                this.attemptsLeft = 0;
+                verifyGameStatus();
+                return false;
+            }
+            return true;
+        } else {
+            System.out.println(String.format("valid coordinates should be between 0 - %s", this.moleGrid.length - 1));
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java]: MINOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+    /**
+     * Method to whack a location in order to search a mole.
+     * @param x x location of grid game.
+     * @param y y location of grid game
+     */
+    void whack(int x, int y) {
+        if (validatedCoordinates(x, y)) {
+            if (this.moleGrid[x][y] == MOLE) {
+                System.out.println("Good, you found a mole!!");
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java]: MINOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            this.attemptsLeft--;
+            verifyGameStatus();
+        }
+    }
+
+    /**
+     * Method to print the grid available for user.
+     */
+    void printGridToUser() {
+        System.out.println(Arrays.deepToString(this.trackerMoleGrid));
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java]: MINOR Replace this use of System.out or System.err by a logger. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS106)

File: src/test/java/org/minions/devfund/damianvp/WhackAMoleTest.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+     */
+    @Test
+    public void testPlaceMoleUsingNegativeLocations() {
+//        WhackAMole game = startNewGame(4,4);
+        final int xLocation = -1;
+        final int yLocation = -1;
+        int expectedMolesLeft = 0;
+        assertFalse(whackAMoleGame.place(xLocation, yLocation));
+        assertEquals(expectedMolesLeft, whackAMoleGame.getMolesLeft());
+//        assertTrue(whackAMoleGame.place());
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/damianvp/WhackAMoleTest.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: src/test/java/org/minions/devfund/damianvp/WhackAMoleTest.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+        assertTrue(whackAMoleGame.place(xLocation, yLocation));
+        assertEquals(expectedMolesLeft, whackAMoleGame.getMolesLeft());
+    }
+
+    /**
+     * Test place Mole using negative locations.
+     */
+    @Test
+    public void testPlaceMoleUsingNegativeLocations() {
+//        WhackAMole game = startNewGame(4,4);
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/damianvp/WhackAMoleTest.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by carledriss** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 8 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+            this.attemptsLeft--;
+            verifyGameStatus();
+        }
+    }
+
+    /**
+     * Method to print the grid available for user.
+     */
+    void printGridToUser() {
+        LOGGER.log(Level.INFO, Arrays.deepToString(this.trackerMoleGrid));
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java]: MAJOR Invoke method(s) only conditionally. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2629)

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    /**
+     * Method to print the complete grid game status.
+     */
+    void printGrid() {
+        LOGGER.log(Level.INFO, Arrays.deepToString(this.moleGrid));
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java]: MAJOR Invoke method(s) only conditionally. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2629)

**Comment by carledriss** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Review by carledriss — CHANGES_REQUESTED** [review]
This PR should only have WackAMole classes

File: src/test/java/org/minions/devfund/damianvp/WhackAMoleTest.java [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+    @Before
+    public void startNewGame() {
+        whackAMoleGame = new WhackAMole(ATTEMPTS, GRID_DIMENSION);
+    }
+
+    /**
+     * Test place Mole using valid locations.
+     */
+    @Test
+    public void testPlaceMoleValidPositions() {
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/damianvp/WhackAMoleTest.java:31]: Please write your unit tests like edx
You should have the same quantity of unit tests

**Review by carledriss — CHANGES_REQUESTED** [review]

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+        int max = this.moleGrid.length;
+        int min = 0;
+        boolean moleIsPlaced;
+        int limitGrid = this.moleGrid.length * this.moleGrid.length;
+        if (moles <= limitGrid) {
+            while (moles > 0) {
+                int x = random.nextInt(max - min);
+                int y = random.nextInt(max - min);
+                moleIsPlaced = place(x, y);
+                if (moleIsPlaced) {
```
Comment (sphinix27) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java:80]: You can remove the Boolean and just use the method, the var is not used elsewhere.
```java
if (place(x,y)) {
     moles--;
}
```

File: src/main/java/org/minions/devfund/damianvp/WhackAMole.java [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+     */
+    boolean validatedCoordinates(int x, int y) {
+        if (x > INVALID_RANGE && y > INVALID_RANGE && x < this.moleGrid.length && y < this.moleGrid.length) {
+            if (x == -1 && y == -1) {
+                this.attemptsLeft = 0;
+                verifyGameStatus();
+                return false;
+            }
+            return true;
+        } else {
```
Comment (sphinix27) [inline src/main/java/org/minions/devfund/damianvp/WhackAMole.java:139]: You can remove the else and just put the Logger and return false after the if condition.

**Review by sphinix27 — CHANGES_REQUESTED** [review]
Increase code coverage.

**Pull request closed without merging**